
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://github.com/jfnavarro/st_pipeline/2.0.0.dev/usage/">
      
      
        <link rel="prev" href="../installation/">
      
      
        <link rel="next" href="../api/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.3">
    
    
      
        <title>Usage - ST Pipeline</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.d7758b05.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="preference" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#usage" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="ST Pipeline" class="md-header__button md-logo" aria-label="ST Pipeline" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            ST Pipeline
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Usage
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/jfnavarro/st_pipeline" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    jfnavarro/st_pipeline
  </div>
</a>
      </div>
    
  </nav>
  
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="ST Pipeline" class="md-nav__button md-logo" aria-label="ST Pipeline" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    ST Pipeline
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/jfnavarro/st_pipeline" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    jfnavarro/st_pipeline
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../installation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Usage
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Usage
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#usage" class="md-nav__link">
    <span class="md-ellipsis">
      Usage
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Usage">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#input-format" class="md-nav__link">
    <span class="md-ellipsis">
      Input format
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#example" class="md-nav__link">
    <span class="md-ellipsis">
      Example
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#visium" class="md-nav__link">
    <span class="md-ellipsis">
      Visium
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#extra-tools" class="md-nav__link">
    <span class="md-ellipsis">
      Extra tools
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Extra tools">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#emsembl-ids" class="md-nav__link">
    <span class="md-ellipsis">
      Emsembl ids
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#merge-demultiplexed-fastq-files" class="md-nav__link">
    <span class="md-ellipsis">
      Merge demultiplexed FASTQ files
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filter-out-genes-by-gene-type" class="md-nav__link">
    <span class="md-ellipsis">
      Filter out genes by gene type
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Filter out genes by gene type">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#remove-spots-from-dataset" class="md-nav__link">
    <span class="md-ellipsis">
      Remove spots from dataset
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#quality-stats" class="md-nav__link">
    <span class="md-ellipsis">
      Quality stats
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../api/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Modules
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../contributing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Contributing
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../changelog/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Changelog
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#usage" class="md-nav__link">
    <span class="md-ellipsis">
      Usage
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Usage">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#input-format" class="md-nav__link">
    <span class="md-ellipsis">
      Input format
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#example" class="md-nav__link">
    <span class="md-ellipsis">
      Example
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#visium" class="md-nav__link">
    <span class="md-ellipsis">
      Visium
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#extra-tools" class="md-nav__link">
    <span class="md-ellipsis">
      Extra tools
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Extra tools">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#emsembl-ids" class="md-nav__link">
    <span class="md-ellipsis">
      Emsembl ids
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#merge-demultiplexed-fastq-files" class="md-nav__link">
    <span class="md-ellipsis">
      Merge demultiplexed FASTQ files
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filter-out-genes-by-gene-type" class="md-nav__link">
    <span class="md-ellipsis">
      Filter out genes by gene type
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Filter out genes by gene type">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#remove-spots-from-dataset" class="md-nav__link">
    <span class="md-ellipsis">
      Remove spots from dataset
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#quality-stats" class="md-nav__link">
    <span class="md-ellipsis">
      Quality stats
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


  <h1>Usage</h1>

<h2 id="usage">Usage<a class="headerlink" href="#usage" title="Permanent link">&para;</a></h2>
<h3 id="input-format">Input format<a class="headerlink" href="#input-format" title="Permanent link">&para;</a></h3>
<p>The following files/parameters are commonly required:</p>
<ul>
<li>FASTQ files (Read 1 containing the barcode and the UMI and read 2 containing the genomic sequence)</li>
<li>A genome index generated with STAR</li>
<li>An annotation file in GTF or GFF3 format (optional when using a transcriptome)</li>
<li>A file containing the barcodes and array coordinates (look at the folder "ids" to use it as a reference).
Basically this file contains 3 columns (BARCODE, X and Y), so if you provide this
file with barcodes identifying cells (for example), the ST pipeline can be used for single cell data.
This file is also optional if the data is not barcoded (for example RNA-Seq data).</li>
<li>A name for the dataset</li>
</ul>
<p>The ST pipeline has multiple parameters mostly related to trimming, mapping, demultiplexing and annotation
but generally the default values are good enough.</p>
<p>The input FASTQ files can be given in gzip/bzip format as well.</p>
<p>Note that the minimum read length is dependant on the type of kit used, and
should be adjusted accordingly, i.e. a 150bp kit should have a different
minimum read length than a 75bp kit.</p>
<p>Soft clipping is also not recommended when using the 75bp kit, due to the shorter length.</p>
<p>The UMI filter can be used for array batches 1000L6 and earlier. It is
not recommended to use it for array batches 1000L7 and newer as the UMI in
these arrays is fully randomised.</p>
<p>The output of the ST Pipeline is a counts matrix (TSV) and a log file with status and useful information.</p>
<p>You can see a full description of the parameters typing <code>st_pipeline_run --help</code> after you have installed the ST pipeline.</p>
<div class="highlight"><pre><span></span><code><span class="go">st_pipeline_run [options] fastq_file_fw fastq_file_rv</span>

<span class="go">  fastq_file_fw</span>
<span class="go">    Read_1 containing the spatial barcodes and UMIs for each sequence.</span>

<span class="go">  fastq_file_rv</span>
<span class="go">    Read_2 containing the gene sequence corresponding to the sequence in</span>
<span class="go">    Read_1.</span>

<span class="go">  -h, --help            show this help message and exit</span>
<span class="go">  --ids [FILE]          Path to the file containing the map of barcodes to the</span>
<span class="go">                        array coordinates</span>
<span class="go">  --ref-map [FOLDER]    Path to the folder with the STAR index for the genome</span>
<span class="go">                        that you want to use as reference</span>
<span class="go">  --ref-annotation [FILE]</span>
<span class="go">                        Path to the reference annotation file (GTF or GFF</span>
<span class="go">                        format is required) to be used to annotated the mapped</span>
<span class="go">                        reads</span>
<span class="go">  --expName [STRING]    Name of the dataset (The output files will prepend</span>
<span class="go">                        this name)</span>
<span class="go">  --contaminant-index [FOLDER]</span>
<span class="go">                        Path to the folder with a STAR index with a</span>
<span class="go">                        contaminant genome reference. Reads will be filtered</span>
<span class="go">                        using the specified genome and mapping reads will be</span>
<span class="go">                        discarded</span>
<span class="go">  --no-clean-up         Do not remove temporary/intermediary files (useful for</span>
<span class="go">                        debugging)</span>
<span class="go">  --verbose             Show extra information on the log file</span>
<span class="go">  --threads [INT]       Number of threads to use (default: 4)</span>
<span class="go">  --bin-path [FOLDER]   Path to folder where binary executables are present</span>
<span class="go">                        (system path by default)</span>
<span class="go">  --log-file [STR]      Name of the file that we want to use to store the logs</span>
<span class="go">                        (default output to screen)</span>
<span class="go">  --output-folder [FOLDER]</span>
<span class="go">                        Path of the output folder</span>
<span class="go">  --temp-folder [FOLDER]</span>
<span class="go">                        Path of the location for temporary files</span>
<span class="go">  --keep-discarded-files</span>
<span class="go">                        Keep files with discarded reads in every step</span>
<span class="go">  --qual-64             Use phred-64 quality instead of phred-33(default) in</span>
<span class="go">                        the quality trimming step</span>
<span class="go">  --min-length-qual-trimming [INT]</span>
<span class="go">                        Minimum length of the reads after trimming, shorter</span>
<span class="go">                        reads will be discarded (default: 20)</span>
<span class="go">  --min-quality-trimming [INT]</span>
<span class="go">                        Minimum phred quality a base must have in order to be</span>
<span class="go">                        kept in the quality trimming step (default: 20)</span>
<span class="go">  --remove-polyA [INT]  Remove PolyA stretches of the given length from R2</span>
<span class="go">                        (Use 0 to disable it) (default: 10)</span>
<span class="go">  --remove-polyT [INT]  Remove PolyT stretches of the given length from R2</span>
<span class="go">                        (Use 0 to disable it) (default: 10)</span>
<span class="go">  --remove-polyG [INT]  Remove PolyG stretches of the given length from R2</span>
<span class="go">                        (Use 0 to disable it) (default: 10)</span>
<span class="go">  --remove-polyC [INT]  Remove PolyC stretches of the given length from R2</span>
<span class="go">                        (Use 0 to disable it) (default: 10)</span>
<span class="go">  --remove-polyN [INT]  Remove PolyN stretches of the given length from R2</span>
<span class="go">                        (Use 0 to disable it) (default: 10)</span>
<span class="go">  --homopolymer-mismatches [INT]</span>
<span class="go">                        Number of mismatches allowed when removing</span>
<span class="go">                        homopolymers (A, T, G, C or N) (default: 0)</span>
<span class="go">  --filter-AT-content [INT%]</span>
<span class="go">                        Discards reads whose number of A and T bases in total</span>
<span class="go">                        are more or equal than the percentage given as input</span>
<span class="go">                        (0-100) (default: 90)</span>
<span class="go">  --filter-GC-content [INT%]</span>
<span class="go">                        Discards reads whose number of G and C bases in total</span>
<span class="go">                        are more or equal the percentage given as input</span>
<span class="go">                        (0-100) (default: 90)</span>
<span class="go">  --mapping-rv-trimming [INT]</span>
<span class="go">                        Number of bases to trim in the reverse reads (R2) for</span>
<span class="go">                        the mapping step (5&#39; end) (default: 0)</span>
<span class="go">  --inverse-mapping-rv-trimming [INT]</span>
<span class="go">                        Number of bases to trim in the reverse reads (R2) for</span>
<span class="go">                        the mapping step (3&#39; end) (default: 0)</span>
<span class="go">  --disable-multimap    If activated, multiple aligned reads obtained during</span>
<span class="go">                        mapping will be all discarded. Otherwise the highest</span>
<span class="go">                        scored one will be kept</span>
<span class="go">  --disable-clipping    If activated, disable soft-clipping (local alignment)</span>
<span class="go">                        in the mapping step</span>
<span class="go">  --min-intron-size [INT]</span>
<span class="go">                        Minimum allowed intron size when searching for splice</span>
<span class="go">                        variants with STAR Splices alignments are disabled by</span>
<span class="go">                        default (=1) but to turn it on set this parameter to a</span>
<span class="go">                        bigger number, for example 10 or 20. (default: 1)</span>
<span class="go">  --max-intron-size [INT]</span>
<span class="go">                        Maximum allowed intron size when searching for splice</span>
<span class="go">                        variants with STAR Splices alignments are disabled by</span>
<span class="go">                        default (=1) but to turn it on set this parameter to a</span>
<span class="go">                        big number, for example 10000 or 100000. (default: 1)</span>
<span class="go">  --star-two-pass-mode  Activates the 2-pass mode in STAR to improve mapping</span>
<span class="go">                        accuracy</span>
<span class="go">  --star-genome-loading [STRING]</span>
<span class="go">                        Similar to the STAR option --genomeLoad. It allows to</span>
<span class="go">                        load the genome index into memory so it can easily be</span>
<span class="go">                        shared by other jobs to save loading time. Read the</span>
<span class="go">                        STAR manual for more info on this. (default:</span>
<span class="go">                        NoSharedMemory)</span>
<span class="go">  --star-sort-mem-limit STAR_SORT_MEM_LIMIT</span>
<span class="go">                        The maximum available RAM for sorting BAM during</span>
<span class="go">                        mapping with STAR. Default is 0 which means that it</span>
<span class="go">                        will be set to the genome index size</span>
<span class="go">  --demultiplexing-mismatches [INT]</span>
<span class="go">                        Number of allowed mismatches when demultiplexing the</span>
<span class="go">                        reads against the barcodes with TaggD (default: 2)</span>
<span class="go">  --demultiplexing-kmer [INT]</span>
<span class="go">                        KMer size to use when demultiplexing against the</span>
<span class="go">                        barcodes with TaggD (default: 6)</span>
<span class="go">  --demultiplexing-overhang [INT]</span>
<span class="go">                        Extra flanking bases added on each side of the barcode</span>
<span class="go">                        when demultiplexing against the barcodes with TaggD</span>
<span class="go">                        (default: 0)</span>
<span class="go">  --demultiplexing-start [INT]</span>
<span class="go">                        Start position of the IDs (Barcodes) in R1 (counting</span>
<span class="go">                        from 0) (default: 0)</span>
<span class="go">  --demultiplexing-metric [STRING]</span>
<span class="go">                        Distance metric to use for TaggD demultiplexing:</span>
<span class="go">                        Options: Subglobal, Levenshtein or Hamming (default:</span>
<span class="go">                        Subglobal)</span>
<span class="go">  --demultiplexing-multiple-hits-keep-one</span>
<span class="go">                        When multiple ambiguous hits with same score are found</span>
<span class="go">                        in the demultiplexing step, keep only one (random).</span>
<span class="go">  --demultiplexing-trim-sequences DEMULTIPLEXING_TRIM_SEQUENCES</span>
<span class="go">                        Trim the barcodes in the input file when doing</span>
<span class="go">                        demultiplexing. The input given is a list of tuples</span>
<span class="go">                        START END START END where START is the integer</span>
<span class="go">                        position of the first base (0 based) and END is the</span>
<span class="go">                        integer position of the last base (1 based). The final</span>
<span class="go">                        barcode will be obtained by combining all the</span>
<span class="go">                        sequences given in the input. This is useful when</span>
<span class="go">                        having a barcode composed of multiple sequences in the</span>
<span class="go">                        reador when the barcode needs to be trimmed out.</span>
<span class="go">                        Trimmng sequences can be given several times.</span>
<span class="go">  --demultiplexing-chunk-size [INT]</span>
<span class="go">                        Chunk size for parallel processing (number of reads assigned to each thread) (default: 10000)</span>
<span class="go">  --htseq-mode [STRING]</span>
<span class="go">                        Mode of annotation when using htseq-count. Modes =</span>
<span class="go">                        {union, intersection-nonempty(default), intersection-</span>
<span class="go">                        strict}</span>
<span class="go">  --htseq-no-ambiguous  When using htseq-count discard reads annotating</span>
<span class="go">                        ambiguous genes (default False)</span>
<span class="go">  --htseq-features HTSEQ_FEATURES [HTSEQ_FEATURES ...]</span>
<span class="go">                        Which feature types to use from the GTF/GFF file in the annotation.</span>
<span class="go">                        Can be given more than one type (default exon)</span>
<span class="go">  --strandness [STRING]</span>
<span class="go">                        What strandness mode to use when annotating with</span>
<span class="go">                        htseq-count [no, yes(default), reverse]</span>
<span class="go">  --include-non-annotated</span>
<span class="go">                        Do not discard un-annotated reads (they will be</span>
<span class="go">                        labeled __no_feature)</span>
<span class="go">  --umi-cluster-algorithm [STRING]</span>
<span class="go">                        Type of clustering algorithm to use when performing</span>
<span class="go">                        UMIs duplicates removal. Options = {naive,</span>
<span class="go">                        hierarchical, Affinity, Adjacent and</span>
<span class="go">                        AdjacentBi(default)} Note that for the affinity method</span>
<span class="go">                        the umi allowed mismatches parameter will be ignored.</span>
<span class="go">  --umi-allowed-mismatches [INT]</span>
<span class="go">                        Number of allowed mismatches (hamming distance) that</span>
<span class="go">                        UMIs of the same gene-spot must have in order to</span>
<span class="go">                        cluster together (default: 1)</span>
<span class="go">  --umi-start-position [INT]</span>
<span class="go">                        Position in R1 (base wise) of the first base of the</span>
<span class="go">                        UMI (starting by 0) (default: 18)</span>
<span class="go">  --umi-end-position [INT]</span>
<span class="go">                        Position in R1 (base wise) of the last base of the UMI</span>
<span class="go">                        (starting by 1) (default: 27)</span>
<span class="go">  --umi-filter          Enables the UMI quality filter based on the template</span>
<span class="go">                        given in --umi-filter-template</span>
<span class="go">  --umi-filter-template [STRING]</span>
<span class="go">                        UMI template (IUPAC nucleotide code) for the UMI</span>
<span class="go">                        filter, default = WSNNWSNNV</span>
<span class="go">  --umi-quality-bases [INT]</span>
<span class="go">                        Maximum number of low quality bases allowed in an UMI</span>
<span class="go">                        (default: 6)</span>
<span class="go">  --umi-counting-offset [INT]</span>
<span class="go">                        UMI count for each gene-spot combination is computed</span>
<span class="go">                        as the number of unique UMIs in each strand/start</span>
<span class="go">                        position. However some reads might have slightly</span>
<span class="go">                        different start positions due to amplification</span>
<span class="go">                        artifacts. This parameters allows to define an offset</span>
<span class="go">                        window from where to count unique UMIs. You can set it</span>
<span class="go">                        to a very high value +9999 to count unique UMIs for</span>
<span class="go">                        the whole gene (default: 250)</span>
<span class="go">  --compute-saturation  Performs a saturation curve computation by sub-</span>
<span class="go">                        sampling the annotated reads, computing unique UMIs</span>
<span class="go">                        and adding the stats to the log file (this can be used</span>
<span class="go">                        to plot saturation curves)</span>
<span class="go">  --saturation-points SATURATION_POINTS [SATURATION_POINTS ...]</span>
<span class="go">                        Saturation points for the saturation curve computation</span>
<span class="go">                        can be provided instead of using default values.</span>
<span class="go">                        Provide a list of values like this for example: 10000</span>
<span class="go">                        20000 50000 100000</span>
<span class="go">  --disable-trimming    Use this flag if you want to skip the trimming step</span>
<span class="go">  --disable-mapping     Use this flag if you want to skip the mapping step. Requires that a &#39;mapped.bam&#39; file is present in --temp-folder.</span>
<span class="go">  --disable-annotation  Use this flag if you want to skip the annotation</span>
<span class="go">  --disable-barcode     Use this flag if you want to skip the barcode</span>
<span class="go">                        demultiplexing step</span>
<span class="go">  --disable-umi         Use this flag if you want to skip the UMI filtering</span>
<span class="go">                        step</span>
<span class="go">  --transcriptome       Use this flag if you want to use transcriptome instead</span>
<span class="go">                        of a genome, the gene tag will be obtained from the</span>
<span class="go">                        transcriptome file</span>
<span class="go">  --version             show program&#39;s version number and exit</span>
</code></pre></div>
<h3 id="example">Example<a class="headerlink" href="#example" title="Permanent link">&para;</a></h3>
<p>An example run would be:</p>
<div class="highlight"><pre><span></span><code><span class="go">st_pipeline_run --expName test --ids ids_file.txt \</span>
<span class="go">  --ref-map path_to_index --htseq-no-ambiguous --log-file log_file.txt --output-folder /home/me/results \</span>
<span class="go">  --ref-annotation annotation_file.gtf --contaminant-index path_to_cont_index file1.fastq file2.fastq</span>
</code></pre></div>
<h3 id="visium">Visium<a class="headerlink" href="#visium" title="Permanent link">&para;</a></h3>
<p>To process Visium datasets it is recommended to use these options:</p>
<div class="highlight"><pre><span></span><code><span class="go">--demultiplexing-mismatches 1</span>
<span class="go">--demultiplexing-kmer 4</span>
<span class="go">--umi-allowed-mismatches 2</span>
<span class="go">--umi-start-position 16</span>
<span class="go">--umi-end-position 28</span>
</code></pre></div>
<h3 id="extra-tools">Extra tools<a class="headerlink" href="#extra-tools" title="Permanent link">&para;</a></h3>
<p>The ST Pipeline ships many scripts that will be installed automatically and that
can be very useful to pre/post process the data and general QC stats and plots.</p>
<h4 id="emsembl-ids">Emsembl ids<a class="headerlink" href="#emsembl-ids" title="Permanent link">&para;</a></h4>
<p>If you used an Ensembl annotation file and you would like change
the output file so it contains gene ids/names instead of Ensembl ids.
You can use the script <code>convertEnsemblToNames</code> that comes with the ST Pipeline</p>
<div class="highlight"><pre><span></span><code><span class="go">convertEnsemblToNames --annotation path_to_annotation_file --output st_data_updated.tsv st_data.tsv</span>
</code></pre></div>
<h4 id="merge-demultiplexed-fastq-files">Merge demultiplexed FASTQ files<a class="headerlink" href="#merge-demultiplexed-fastq-files" title="Permanent link">&para;</a></h4>
<p>If you used different indexes to sequence and need to merge the fastq files
you can use the script <code>merge_fastq</code> that comes with the ST Pipeline</p>
<div class="highlight"><pre><span></span><code><span class="go">merge_fastq --run-path path_to_run_folder --out-path path_to_output --identifiers S1 S2 S3 S4</span>
</code></pre></div>
<p>Where <code>--identifiers</code> will be strings that identify each demultiplexed sample.</p>
<h3 id="filter-out-genes-by-gene-type">Filter out genes by gene type<a class="headerlink" href="#filter-out-genes-by-gene-type" title="Permanent link">&para;</a></h3>
<p>If you want to remove from the dataset (matrix in TSV) genes corresponding
to certain gene types (For instance to keep only protein_coding). You can do
so with the script <code>filter_gene_type_matrix</code> that comes with the ST Pipeline</p>
<div class="highlight"><pre><span></span><code><span class="go">filter_gene_type_matrix --gene-types-keep protein-coding --annotation path_to_annotation_file stdata.tsv</span>
</code></pre></div>
<p>You may include the parameter <code>--ensembl-ids</code> if your genes are represented as Emsembl ids instead.</p>
<p>The value of <code>--gene-types-keep</code> must match the annotation file provided.</p>
<h4 id="remove-spots-from-dataset">Remove spots from dataset<a class="headerlink" href="#remove-spots-from-dataset" title="Permanent link">&para;</a></h4>
<p>If you want to remove spots from a dataset (matrix in TSV) for instance
to keep only spots inside the tissue. You can do so with the script <code>adjust_matrix_coordinates</code>
that comes with the ST Pipeline</p>
<div class="highlight"><pre><span></span><code><span class="go">adjust_matrix_coordinates --outfile new_stdata.tsv --coordinates-file coordinates.txt stdata.tsv</span>
</code></pre></div>
<p>Where <code>coordinates.txt</code> must be a tab delimited file with 6 columns:</p>
<div class="highlight"><pre><span></span><code><span class="go">orig_x orig_y new_x new_y new_pixel_x new_pixel_y</span>
</code></pre></div>
<p>Only spots whose coordinates in the file will be kept and then optionally you
can update the coordinates in the matrix choosing for the new array or pixel coordinates.</p>
<h4 id="quality-stats">Quality stats<a class="headerlink" href="#quality-stats" title="Permanent link">&para;</a></h4>
<p>The ST Pipeline generate useful stats/QC information in the LOG file but if you
want to obtain more detailed information about the quality of the data, you can run the following script:</p>
<div class="highlight"><pre><span></span><code><span class="go">st_qa stdata.tsv</span>
</code></pre></div>
<p>If you want to perform quality stats on multiple datasets you can run:</p>
<div class="highlight"><pre><span></span><code><span class="go">multi_qa stdata1.tsv stadata2.tsv stdata3.tsv stdata4.tsv</span>
</code></pre></div>
<p><code>multi_qa</code> generates violing plots, correlation plots/tables and more useful information and
it allows to log the counts for the correlation analaysis.</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
    <a href="https://github.com/jfnavarro/st_pipeline" target="_blank" rel="noopener" title="Github" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
    </a>
  
    
    
    
    
    <a href="mailto:jc.fernandez.navarro@gmail.com" target="_blank" rel="noopener" title="" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m20 8-8 5-8-5V6l8 5 8-5m0-2H4c-1.11 0-2 .89-2 2v12a2 2 0 0 0 2 2h16a2 2 0 0 0 2-2V6a2 2 0 0 0-2-2"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": ["navigation.indexes", "navigation.instant", "navigation.tabs.sticky"], "search": "../assets/javascripts/workers/search.f8cc74c7.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.f1b6f286.min.js"></script>
      
    
  </body>
</html>